core/packages/contracts/src/NativeTokens.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import "openzeppelin/access/Ownable.sol";
+import "openzeppelin/access/AccessControl.sol";
+import "openzeppelin/token/ERC20/extensions/IERC20Metadata.sol";
+
+import "./TokenVault.sol";
+import "./SubstrateTypes.sol";
+import "./NativeTokensTypes.sol";
+import "./OutboundChannel.sol";
+
+/// @title Native Tokens
+/// @dev Manages locking, unlocking ERC20 tokens in the vault. Initializes ethereum native
+/// tokens on the substrate side via create.
+contract NativeTokens is AccessControl {
+    /// @dev Describes the type of message.
+    enum Action {
+        Unlock
+    }
+
+    /// @dev Message format.
+    struct Message {
+        /// @dev The action type.
+        Action action;
+        /// @dev The message payload.
+        bytes payload;
+    }
+
+    /// @dev Unlock payload format.
+    struct UnlockPayload {
+        /// @dev The ERC20 token to unlock.
+        address token;
+        /// @dev The destination address that will receive unlocked funds.
+        address recipient;
+        /// @dev The amount to unlock.
+        uint128 amount;
+    }
+
+    /// @dev Emitted once the funds are locked and a message is successfully queued.
+    event Locked(bytes recipient, address token, uint128 amount);
+
+    /// @dev Emitted once the funds are unlocked.
+    event Unlocked(address recipient, address token, uint128 amount);
+
+    /// @dev Emitted after enqueueing a a create token message to substrate.
+    event Created(address token);
+
+    /// @dev Set a new outbound channel.
+    event OutboundChannelUpdated(address newOutBoundChannel);
+
+    /* State */
+
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+    bytes32 public constant SENDER_ROLE = keccak256("SENDER_ROLE");
+
+    bytes32 public immutable peerID;
+    bytes public peer;
```

Comment:
comment to define/explain these 2 variables I think needed. Or maybe could also have some kinda PeeredApp base class that includes these fields which NativeTokens inherits from and which defines these concepts more clearly? PeeredApp base class could also be useful for future devs when we become permissionless?